### PR TITLE
WIP custom metrics

### LINKED ISSUE #292: Accuracy and metrics with Model
In #286 I briefly talk about the idea of separating the metrics computation (like the accuracy) from `Model`. At the moment, you can keep track of the accuracy in the logs (both history and console logs) easily with the flag `show_accuracy=True` in `Model.fit()`. Unfortunately this is limited to the accuracy and does not handle any other metrics that could be valuable to the user.

We could have the computation of these metrics outside of `Model` and call them with callbacks if one wants to keep track of them during training. It may be valuable for the future but for it could also raise some issues short term
- It would be impossible to log the accuracy (or any other metrics) with the base logger as callbacks do not interact with each other. One solution would be to let the user create her own logger on a different level of verbosity (possibly by inheriting from the current `BaseLogger`).
- We would have to think about how to access the training and validation set with callbacks.

### PR DESCRIPTION
**Pull request by stevenxxiu**
This is my attempt to try and resolve #292. The gist of it is to allow custom metrics being input in a dict format into `fit`, with the keys being the labels, e.g.:

    {'acc': accuracy_metric}

I haven't tested it yet, but does it look like something reasonable?

### CODE REVIEW

File: keras/models.py [hunk 1]
```diff
@@ -175,6 +175,15 @@ def get_function_name(o):
         return o.__name__
 
 
+def accuracy_metric(y, y_gold, class_mode):
```
Comment (fchollet) [inline keras/models.py]: The variable naming convention used elsewhere in the code base is `y_pred`, `y_true`.

**Comment by fchollet** [comment]
> I haven't tested it yet, but does it look like something reasonable?

It's not working yet, but yes, it looks interesting to me. 

Some more discussion would be necessary with regard to what interface should be used for configuring the metrics (I don't think a dictionary of functions is very clean. We should probably follow a similar approach as what is being done with activations and losses: being able to pass either strings or Keras functions. Metric names could be provided as part of a tuple (name, function) or could be the function name).

**Comment by stevenxxiu** [comment]
I've updated it so that metrics are provided as a list, and names are obtained via `__name__`. Validation names are prefixed with `val_`.

File: keras/models.py [hunk 2]
```diff
@@ -563,17 +552,14 @@ class accuracy in the logs to stdout at each epoch.
...
-            f = self._train
-            out_labels = ['loss']
+        f = self._train
+        out_labels = ['loss'] + [metric.__name__ for metric in self.train_metrics]
 
         sample_weight = standardize_weights(y, class_weight=class_weight,
                                             sample_weight=sample_weight)
         ins = X + [y, sample_weight]
-        metrics = ['loss', 'acc', 'val_loss', 'val_acc']
+        metrics = ['loss'] + [metric.__name__ for metric in self.train_metrics] + \
```
Comment (fchollet) [inline keras/models.py]: `\` is to be avoided. This line can be rewritten as several successive statements.

File: keras/metrics.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+from __future__ import absolute_import
+from . import backend as K
+
+
+def accuracy(y_pred, y_true, class_mode):
```
Comment (fchollet) [inline keras/metrics.py]: If we are going to use `__name__` as the display name of a metric, then this should be renamed "acc" to match expectations in the existing codebase. But another way may be possible. How about some polymorphisme: elements in the metrics array can be either functions, strings (an alias to a default function), or tuples. When tuple, it is expected that the first element is a string (the display name of the metric) and the second is a function or a string alias of a function.

**Comment by stevenxxiu** [comment]
I did a bit more thinking on this, and some metrics such as f1 don't have averaging as aggregation. Perhaps each metric can be a class instead? I've come up with the following for f1:

    class f1:
        def __init__(num_classes, classes):
            self.num_classes = num_classes
            self.classes = classes

        def batch_statistics(self):
            if class_mode == 'categorical':
                y_pred = T.argmax(y_pred, axis=-1)
                if y_true.ndim == 2:
                    y_true = T.argmax(y_true, axis=-1)
                return {
                    'tp_sum': T.extra_ops.bincount(y_true, T.eq(y_true, y_pred), minlength=self.num_classes)[self.classes]
                    'pred_sum': T.extra_ops.bincount(y_pred, minlength=self.num_classes)[self.classes]
                    'true_sum': T.extra_ops.bincount(y_true, minlength=self.num_classes)[self.classes]
                }
            elif class_mode == 'binary':
                raise NotImplementedError
            else:
                raise Exception('Invalid class mode:' + str(class_mode))

        def aggregate(self, batches):
            tp_sum = np.mean(batches['tp_sum'])
            pred_sum = np.mean(batches['pred_sum'])
            true_sum = np.mean(batches['true_sum'])
            precision = tp_sum / pred_sum
            recall = tp_sum / true_sum
            f_score = 2 * precision * recall / (precision + recall)
            return np.mean(f_score)

**Comment by fchollet** [comment]
Closing outdated PRs following the release of Keras 1.0.
